### ISSUE TITLE
Memory usage for aotudiff difference with different backend

### ISSUE THREAD

**Issue by IanYangChina** [taichi-dev/taichi#8420]
Hi!

I have researched on this topic for a while.
I noticed a difference on how GPU memory is used when using different backends.
It seems like the cuda backend occupies quite a lot more memory than other backends like Vulkan and opengl.
It would be nice to know why!

Also, it is strange that the same codes run on a machine with 6GB GPU memory without error, while keep throwing CUDA OUT OF MEMORY error message on another machine with 11GB memory while compiling kernels.

Looking forward to your answers!

**Comment by jim19930609** [comment]
> Also, it is strange that the same codes run on a machine with 6GB GPU memory without error, while keep throwing CUDA OUT OF MEMORY error message on another machine with 11GB memory while compiling kernels.

Usually this is caused by diffs in the environmental variables across the two machines, or are you running some other program on GPU in the meantime?

**Comment by jim19930609** [comment]
> I noticed a difference on how GPU memory is used when using different backends.

Have you set `device_memory_GB` in the `ti.init()`? That will cause preallocation if you have use any Sparse data structure in your code.

**Comment by IanYangChina** [comment]
Hi @jim19930609, thanks for the relies.
No, I did not run other programs.
Isn't the `device_memory` argument only works for `ti.cuda`? However, I leave the `device_memory` the same and use different backends, which still result in different memory usages.

**Comment by jim19930609** [comment]
Yes, the `device_memory` actually only works for CUDA backend. The thing is Vulkan and other backends have their built-in memory pool & allocators so Taichi doesn't have to maintain one.

**Comment by jim19930609** [comment]
So I'd suggest reduce the `device_memory` and see if that takes an effect

**Comment by IanYangChina** [comment]
> So I'd suggest reduce the `device_memory` and see if that takes any effect

It did not have any effect, it looks like the memory exploded when compiling backward kernels in autodiff mode.
It exploded in one machine and didn't in another.

**Comment by IanYangChina** [comment]
It seems like the compilation of backward kernels use different amount of GPU memory when using different backends.

BTW, I noticed that I couldn't select my cuda device when using OpenGL backend, there isn't a global variable for it.
It would be nice to know how to do it.
Thanks for all your answers!
